Validate YAML on Bitbucket Server & Refactor the Validation

tricktron opened the pull request:
# Changes <!-- 🎉🎉🎉 Thank you for the PR!!! 🎉🎉🎉 -->

<!-- Describe your changes here- ideally you can get that description straight from
your descriptive commit message(s)! -->

The main goal of this pull request is to better surface error logs to the pipeline as code developer on Bitbucket Server: https://issues.redhat.com/browse/SRVKP-6767.

Essentially, this pull request does three things in three commits:

- Validates the YAML and prints a friendly message for better debugging purposes on bitbucket server as well. Bitbucket Server was not included in https://github.com/openshift-pipelines/pipelines-as-code/pull/1578.
- Refactores the whole YAML validation to its own function and uses it in all providers.
- Adds the bad YAML test for gitea as well so that we have tested it on all providers.

# Submitter Checklist

- [x] 📝 Please ensure your commit message is clear and informative. For guidance on crafting effective commit messages, refer to the How to write a git commit message guide. We prefer the commit message to be included in the PR body itself rather than a link to an external website (ie: Jira ticket).

- [ ] ♽ Before submitting a PR, run make test lint to avoid unnecessary CI processing. For an even more efficient workflow, consider installing [pre-commit](https://pre-commit.com/) and running pre-commit install in the root of this repository.

- [ ] ✨ We use linters to maintain clean and consistent code. Please ensure you've run make lint before submitting a PR. Some linters offer a --fix mode, which can be executed with the command make fix-linters (ensure [markdownlint](https://github.com/DavidAnson/markdownlint) and [golangci-lint](https://github.com/golangci/golangci-lint) tools are installed first).

- [ ] 📖 If you're introducing a user-facing feature or changing existing behavior, please ensure it's properly documented.

- [x] 🧪 While 100% coverage isn't a requirement, we encourage unit tests for any code changes where possible.

- [ ] 🎁 If feasible, please check if an end-to-end test can be added. See [README](https://github.com/openshift-pipelines/pipelines-as-code/blob/main/test/README.md) for more details.

- [x] 🔎 If there's any flakiness in the CI tests, don't necessarily ignore it. It's better to address the issue before merging, or provide a valid reason to bypass it if fixing isn't possible (e.g., token rate limitations).

tricktron committed:
fix bitbucketserver: validate yaml and propagate

error.

tricktron committed:
refactor: extract yaml validation to its own

function.

tricktron committed:
feat: add bad yaml test to gitea

chmouel commented:
/ok-to-test

chmouel commented:
thanks for the contribution!

cc @zakisk

zakisk commented on pkg/provider/provider.go:
```diff
@@ -123,3 +125,11 @@ func CompareHostOfURLS(uri1, uri2 string) bool {
+
+func ValidateYaml(content []byte, filename string) error {
+	var validYaml any
+	if err := yaml.Unmarshal(content, &validYaml); err != nil {
```
@tricktron why did you choose "k8s.io/apimachinery/pkg/util/yaml" over "gopkg.in/yaml.v2"?, is there any specific reason?

tricktron replied:
> @tricktron why did you choose "k8s.io/apimachinery/pkg/util/yaml" over "gopkg.in/yaml.v2"?, is there any specific reason?

@zakisk Ups, I just reused the `k8s.io/apimachinery/pkg/util/yaml` during the refactoring from [this pr](https://github.com/openshift-pipelines/pipelines-as-code/pull/1578/files#diff-5b6abe4f3dd00e71b0bce9400ca69023651eda401e8b9149734dbe65d79f66dcR25). But now that you mention it, I see that `k8s.io/apimachinery/pkg/util/yaml` is only used once in `provider/github` and the `gopkg.in/yaml.v2` dependency is used more often.

Therfore, it might make sense to only use `gopkg.in/yaml.v2` so that we can ditch `k8s.io/apimachinery/pkg/util/yaml`. What do you think?

tricktron replied:
@zakisk I switched from `k8s.io/apimachinery/pkg/util/yaml` to `gopkg.in/yaml.v2` for yaml validation.

chmouel replied:
i think if i rem ok we need gopkg.yaml for yaml error validation in that pacakge, all other we use k8 machinery yaml because it remove a lot of extra stuff when reading/writting a yaml

i think the e2e test fails on that

zakisk replied:
@tricktron can you please take at [this line](https://github.com/openshift-pipelines/pipelines-as-code/actions/runs/12177814138/job/33968434297?pr=1846#step:11:2024), this is the reason E2E is failing.

tricktron committed:
refactor: use gopkg.in/yaml.v2 for validation

instead of k8s.io/apimachinery/pkg/util/yaml.

zakisk commented:
/test go-testing

chmouel commented:
/retest

tricktron commented:
@zakisk I don´t know why the e2e tests failed (I have not changed them). I rebased on main to get the latest commits and pushed again to run the tests.

zakisk commented:
> @zakisk I don´t know why the e2e tests failed (I have not changed them). I rebased on main to get the latest commits and pushed again to run the tests.

it happens sometimes, let's wait for this new run!

zakisk commented:
@tricktron oh, I missed @chmouel 's comment.

tricktron committed:
fix e2e test: change expected bad yaml string to

match wrapped errors of gopkg.in/yaml.v2 instead of errors of
k8s.io/apimachinery/pkg/util/yaml.

tricktron commented:
@zakisk @chmouel Alright, I also had to fix the `test/testdata/TestGithubPullRequestSecondBadYaml.golden` file. Now all the e2e tests pass as well.

chmouel commented:
Thank you for your contribution ! @tricktron
